### ISSUE TITLE
Best practice for authorizing based on parent model? (i.e. adding context?)

### ISSUE THREAD

**Issue by andrewhavens** [elabs/pundit#403]
I don't see this mentioned in the README, and I'd like to add it if I can get an idea of what the best practice for this is...

Let's say we have a `Blog` model that `has_many :posts`. We only want to allow the "blog owner" to create posts. How would you define and call the policy for this action?

```ruby
class PostPolicy < Struct.new(:user, :post)
  def create?
    return true if blog.owner_id == user.id # how to get access to blog or even blog_id?
  end
end

class PostsController
  def create
    # we have access to params[:blog_id]
    authorize Post # this won't work because authorization is based on the blog model
    # ...
  end
end
```

**Comment by ce07c3** [comment]
Why not authorize on the instance?

```ruby
class PostPolicy < Struct.new(:user, :post)
  def create?
    blog = post.blog

    blog.owner_id == user.id
  end

  def permitted_attributes_for_create
    [:blog_id]
  end
end
```

```ruby
class PostsController
  def create
    post = Post.new(permitted_attributes(Post))
    authorize(post)
    # ...
  end
end
```

**Comment by jnicklas** [comment]
What @ce07c3 said :)

**Comment by andrewhavens** [comment]
@ce07c3 @jnicklas Thanks for your reply. The reason that I'm still struggling with this idea is because I'm also trying to use a service object to perform the majority of the business logic. I feel like there is some duplication here, or it just feels a little awkward. It feels like the service object should be in charge of knowing how to initialize the model. Do you have any suggestions on how to improve this?

```ruby
class PostsController
  def create
    @post = current_user.posts.new(permitted_attributes(Post))
    authorize(@post)
    result = Posts::Create.new(@post).perform
    if result.success?
      render json: result.post, status: :created
    else
      # ...
    end
  end
end

# tl;dr example service 
class Posts::Create < Struct.new(:post)
  def perform
    save_post && other_stuff && more_stuff
    result_object
  end

  def save_post
    return true if post.save
    @errors = post.errors.full_messages
    return false
  end

  # ...
end
```

**Comment by brendon** [comment]
What if we wanted to authorise the `index` action based on the parent model instance. In this case we could only supply the child class rather than an instance.

I've worked around this by: `authorize parent_instance, :manage?` where `manage?` is a special method on the parent's policy. Have I missed something, or is this the right way to go?
